ext/pcre/php_pcre.c

```diff
@@ -2939,6 +2947,108 @@ static PHP_FUNCTION(preg_last_error)
 }
 /* }}} */
 
+static int callout_func(pcre2_callout_block *block, void *user_data) {
+	const char *subject = (const char *) block->subject;
+	PCRE2_SIZE *ovector = block->offset_vector;
+	zval info, matches, retval;
+	uint32_t i;
+	(void) user_data;
+
+	array_init(&info);
+	if (block->callout_string) {
+		add_assoc_string(&info, "callout", (char *) block->callout_string);
+	} else {
+		add_assoc_long(&info, "callout", block->callout_number);
+	}
+
+	add_assoc_stringl(&info, "subject", subject, block->subject_length);
+
+	array_init_size(&matches, block->capture_top);
+	add_assoc_zval(&info, "captures", &matches);
+	for (i = 0; i < block->capture_top; i++) {
+		add_offset_pair(
+			&matches, subject, ovector[2*i], ovector[2*i+1], NULL, /* unmatched_as_null */ 1);
+	}
+
+	add_assoc_long(&info, "capture_last", block->capture_last);
+
+	if (block->mark) {
+		add_assoc_string(&info, "mark", (char *) block->mark);
+	}
+
+	add_assoc_long(&info, "start_match", block->start_match);
+	add_assoc_long(&info, "current_position", block->current_position);
+	add_assoc_long(&info, "pattern_position", block->pattern_position);
+	add_assoc_long(&info, "next_item_length", block->next_item_length);
+
+	ZEND_ASSERT(ZEND_FCI_INITIALIZED(PCRE_G(callout_fci)));
+	PCRE_G(callout_fci).retval = &retval;
+	PCRE_G(callout_fci).params = &info;
+	PCRE_G(callout_fci).param_count = 1;
+
+	if (zend_call_function(&PCRE_G(callout_fci), &PCRE_G(callout_fcc)) == FAILURE
+			|| Z_TYPE(retval) == IS_UNDEF) {
+		zval_ptr_dtor(&info);
+		if (!EG(exception)) {
+			php_error_docref(NULL, E_WARNING, "Failed to call callout function");
+		}
+		return PCRE2_ERROR_CALLOUT;
+	}
+	zval_ptr_dtor(&info);
+
+	if (Z_TYPE(retval) != IS_LONG) {
+		php_error_docref(NULL, E_WARNING, "Callout function must return integer");
+		return PCRE2_ERROR_CALLOUT;
+	}
+
+	if (Z_LVAL(retval) == 0) {
+		/* Proceed matching as usual. */
+		return 0;
+	} else if (Z_LVAL(retval) > 0) {
+		/* Fail at current position. */
+		return 1;
+	} else {
+		/* Fail overall match. */
+		return PCRE2_ERROR_NOMATCH;
+	}
+}
+
+/* {{{ proto callable preg_callout_function(callable $new_function) */
+static PHP_FUNCTION(preg_callout_function)
+{
+	zend_fcall_info fci = {0};
+	zend_fcall_info_cache fcc;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|f!", &fci, &fcc) == FAILURE) {
```

Comment:
I'm curious: Why don't you use the `ZEND_PARSE_PARAMETERS_*` macros?

Reply:
This is a rarely used and not performance sensitive API, so using `zend_parse_parameters` avoids unnecessary codesize increase.